torch/csrc/jit/frontend/ir_emitter.cpp

```diff
@@ -1662,7 +1682,31 @@ struct to_ir {
         graph->createStore(x, fv)->insertBefore(false_block->return_node());
       }
 
-      auto unified = unifyTypes(tv->type(), fv->type());
+      // If we already know that the type is a Union, then we want to
+      // allow heterogeneous types to be unified
+      c10::optional<TypePtr> full_type = environment_stack->getType(x);
+      bool default_to_union =
+          full_type && (*full_type)->kind() == UnionType::Kind;
+      auto unified = unifyTypes(
+          tv->type(), fv->type(), /*default_to_union=*/default_to_union);
+
+      // Special logic for if we have a Union-typed value: check if
+      // the Union we got from unifying the `true` and `false` types
+      // is a subtype of the known type of the value. If so, set
+      // `unified` to be the known type to prevent unintentional
+      // narrowing of the Union type. E.g. `Union[T1, T2, T3]` should
+      // not become `Union[T1, T2]` just because its initial definition
+      // is as "`T1` if cond else `T2`". If the Union we got from type
```

Comment:
The Union is guaranteed to be a subtype of its original type, since we only allow redeclaration/redefinition of variables if they subtype the existing variable in the parent frame.